File: packages/python/src/types/bridge.py
from __future__ import annotations

from typing import Any, Literal, TypedDict

PROTOCOL_VERSION = 1

BRIDGE_PROGRESS_METHOD = "bridgeProgress"

BridgeMethod = Literal[
    "probe",
    "register",
    "ping",
    "pong",
    "scrape",
    "scrape_article",
    "scrape_emails",
    "scrape_phones",
    "scrape_urls",
    "scrape_images",
]


class BridgeJsonRpcError(TypedDict, total=False):
    code: int
    message: str
    data: dict[str, Any]


class BridgeRequest(TypedDict, total=False):
    jsonrpc: str
    id: str
    method: str
    params: dict[str, Any]


class BridgeResponse(TypedDict, total=False):
    jsonrpc: str
    id: str
    result: Any
    error: BridgeJsonRpcError


class RegisterParams(TypedDict, total=False):
    protocolVersion: int
    browser: str
    extensionVersion: str
    deviceId: str
    capabilities: list[str]


class BridgeProgressParams(TypedDict, total=False):
    requestId: str
    phase: str
    message: str
    progress: float
    total: float
    data: dict[str, Any]


class BridgeProgressNotification(TypedDict, total=False):
    jsonrpc: str
    method: str
    params: BridgeProgressParams


def is_bridge_response(msg: Any) -> bool:
    if msg is None or not isinstance(msg, dict):
        return False
    m = msg
    if m.get("jsonrpc") != "2.0":
        return False
    if not isinstance(m.get("id"), str):
        return False
    if isinstance(m.get("method"), str):
        return False
    return m.get("result") is not None or m.get("error") is not None


def is_extension_bridge_request(msg: Any) -> bool:
    if msg is None or not isinstance(msg, dict):
        return False
    m = msg
    if m.get("jsonrpc") != "2.0":
        return False
    if not isinstance(m.get("id"), str):
        return False
    return isinstance(m.get("method"), str)


def is_bridge_progress_notification(msg: Any) -> bool:
    if msg is None or not isinstance(msg, dict):
        return False
    m = msg
    if m.get("jsonrpc") != "2.0":
        return False
    if m.get("method") != BRIDGE_PROGRESS_METHOD:
        return False
    if m.get("id") is not None:
        return False
    return True

File: packages/node/src/types/bridge.ts
export const PROTOCOL_VERSION = 1;

export const BRIDGE_PROGRESS_METHOD = 'bridgeProgress' as const;

export type BridgeMethod =
  | 'probe'
  | 'register'
  | 'ping'
  | 'pong'
  | 'scrape'
  | 'scrape_article'
  | 'scrape_emails'
  | 'scrape_phones'
  | 'scrape_urls'
  | 'scrape_images';

export interface BridgeRequest {
  jsonrpc: '2.0';
  id: string;
  method: BridgeMethod;
  params?: Record<string, unknown>;
}

export interface BridgeResponse {
  jsonrpc: '2.0';
  id: string;
  result?: unknown;
  error?: BridgeJsonRpcError;
}

export interface BridgeJsonRpcError {
  code: number;
  message: string;
  data?: {
    lionscraperCode?: string;
    [key: string]: unknown;
  };
}

export interface RegisterParams {
  protocolVersion: number;
  browser: string;
  extensionVersion: string;
  deviceId: string;
  capabilities: string[];
}

/** Extension → Server JSON-RPC Notification (no `id`). See docs/mcp.md §4.6. */
export interface BridgeProgressParams {
  requestId: string;
  phase?: string;
  message?: string;
  progress?: number;
  total?: number;
  data?: Record<string, unknown>;
}

export type BridgeProgressNotificationMethod = typeof BRIDGE_PROGRESS_METHOD;

export interface BridgeProgressNotification {
  jsonrpc: '2.0';
  method: BridgeProgressNotificationMethod;
  params?: BridgeProgressParams;
}

/** Messages Extension may send after register (response or progress notification). */
export type BridgeExtensionOutboundMessage = BridgeResponse | BridgeProgressNotification;

export type BridgeMessage = BridgeRequest | BridgeResponse;

export function isBridgeRequest(msg: BridgeMessage): msg is BridgeRequest {
  return 'method' in msg && 'id' in msg && (msg as BridgeRequest).id !== undefined && (msg as BridgeRequest).id !== null;
}

export function isBridgeResponse(msg: unknown): msg is BridgeResponse {
  if (msg === null || typeof msg !== 'object') return false;
  const m = msg as Record<string, unknown>;
  if (m.jsonrpc !== '2.0') return false;
  if (typeof m.id !== 'string') return false;
  if (typeof m.method === 'string') return false;
  return m.result !== undefined || m.error !== undefined;
}

/** Extension → Server request with `id` (e.g. `register`, `ping`). */
export function isExtensionBridgeRequest(msg: unknown): msg is BridgeRequest {
  if (msg === null || typeof msg !== 'object') return false;
  const m = msg as Record<string, unknown>;
  if (m.jsonrpc !== '2.0') return false;
  if (typeof m.id !== 'string') return false;
  return typeof m.method === 'string';
}

export function isBridgeProgressNotification(msg: unknown): msg is BridgeProgressNotification {
  if (msg === null || typeof msg !== 'object') return false;
  const m = msg as Record<string, unknown>;
  if (m.jsonrpc !== '2.0') return false;
  if (m.method !== BRIDGE_PROGRESS_METHOD) return false;
  if ('id' in m && m.id !== undefined && m.id !== null) return false;
  return true;
}
